File: book_catalog.go
# book_catalog.go
/**
 * 📚 Book Catalog – Organize Your Library by Genre (Go Edition)
 * Features: add, edit, delete, search, genre stats, persistence
 */

package main

import (
	"bufio"
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"strconv"
	"strings"
	"time"
	"github.com/google/uuid"
)

// ─── Data Model ─────────────────────────────────────────────────────────────

type Book struct {
	ID     string `json:"id"`
	Title  string `json:"title"`
	Author string `json:"author"`
	Genre  string `json:"genre"`
	Year   int    `json:"year"`
	ISBN   string `json:"isbn"`
}

type CatalogData struct {
	Books []Book `json:"books"`
}

// ─── Colors ──────────────────────────────────────────────────────────────────

const (
	reset  = "\x1b[0m"
	bright = "\x1b[1m"
	dim    = "\x1b[2m"
	red    = "\x1b[31m"
	green  = "\x1b[32m"
	yellow = "\x1b[33m"
	blue   = "\x1b[34m"
	magenta = "\x1b[35m"
	cyan   = "\x1b[36m"
)

func c(str, color string) string {
	return color + str + reset
}

// ─── Catalog Manager ──────────────────────────────────────────────────────

type BookCatalog struct {
	books    []Book
	filePath string
}

func NewBookCatalog() *BookCatalog {
	home, _ := os.UserHomeDir()
	dir := filepath.Join(home, ".book_catalog")
	os.MkdirAll(dir, 0755)
	filePath := filepath.Join(dir, "books.json")
	c := &BookCatalog{filePath: filePath}
	c.load()
	return c
}

func (c *BookCatalog) load() {
	if _, err := os.Stat(c.filePath); os.IsNotExist(err) {
		c.books = []Book{}
		return
	}
	raw, err := os.ReadFile(c.filePath)
	if err != nil {
		c.books = []Book{}
		return
	}
	var data CatalogData
	if err := json.Unmarshal(raw, &data); err != nil {
		c.books = []Book{}
		return
	}
	c.books = data.Books
}

func (c *BookCatalog) save() {
	data := CatalogData{Books: c.books}
	raw, _ := json.MarshalIndent(data, "", "  ")
	os.WriteFile(c.filePath, raw, 0644)
}

func (c *BookCatalog) AddBook(title, author, genre string, year int, isbn string) Book {
	book := Book{
		ID:     uuid.New().String(),
		Title:  title,
		Author: author,
		Genre:  genre,
		Year:   year,
		ISBN:   isbn,
	}
	c.books = append(c.books, book)
	c.save()
	return book
}

func (c *BookCatalog) DeleteBook(id string) bool {
	for i, b := range c.books {
		if b.ID == id {
			c.books = append(c.books[:i], c.books[i+1:]...)
			c.save()
			return true
		}
	}
	return false
}

func (c *BookCatalog) UpdateBook(id string, updates map[string]interface{}) bool {
	for i, b := range c.books {
		if b.ID == id {
			if title, ok := updates["title"].(string); ok {
				c.books[i].Title = title
			}
			if author, ok := updates["author"].(string); ok {
				c.books[i].Author = author
			}
			if genre, ok := updates["genre"].(string); ok {
				c.books[i].Genre = genre
			}
			if year, ok := updates["year"].(int); ok {
				c.books[i].Year = year
			}
			if isbn, ok := updates["isbn"].(string); ok {
				c.books[i].ISBN = isbn
			}
			c.save()
			return true
		}
	}
	return false
}

func (c *BookCatalog) GetBook(id string) *Book {
	for _, b := range c.books {
		if b.ID == id {
			return &b
		}
	}
	return nil
}

func (c *BookCatalog) Search(query, genre, author string, yearMin, yearMax *int) []Book {
	results := c.books
	if query != "" {
		q := strings.ToLower(query)
		results = filterBooks(results, func(b Book) bool {
			return strings.Contains(strings.ToLower(b.Title), q) ||
				strings.Contains(strings.ToLower(b.Author), q) ||
				strings.Contains(strings.ToLower(b.Genre), q)
		})
	}
	if genre != "" {
		g := strings.ToLower(genre)
		results = filterBooks(results, func(b Book) bool {
			return strings.ToLower(b.Genre) == g
		})
	}
	if author != "" {
		a := strings.ToLower(author)
		results = filterBooks(results, func(b Book) bool {
			return strings.Contains(strings.ToLower(b.Author), a)
		})
	}
	if yearMin != nil {
		results = filterBooks(results, func(b Book) bool { return b.Year >= *yearMin })
	}
	if yearMax != nil {
		results = filterBooks(results, func(b Book) bool { return b.Year <= *yearMax })
	}
	return results
}

func filterBooks(books []Book, fn func(Book) bool) []Book {
	var res []Book
	for _, b := range books {
		if fn(b) {
			res = append(res, b)
		}
	}
	return res
}

func (c *BookCatalog) GetGenres() map[string]int {
	m := make(map[string]int)
	for _, b := range c.books {
		m[b.Genre]++
	}
	return m
}

func (c *BookCatalog) GetStats() (total, authors, genres int, minYear, maxYear *int) {
	total = len(c.books)
	authorSet := make(map[string]bool)
	for _, b := range c.books {
		authorSet[b.Author] = true
	}
	authors = len(authorSet)
	genres = len(c.GetGenres())
	if total > 0 {
		minY, maxY := c.books[0].Year, c.books[0].Year
		for _, b := range c.books {
			if b.Year < minY {
				minY = b.Year
			}
			if b.Year > maxY {
				maxY = b.Year
			}
		}
		minYear = &minY
		maxYear = &maxY
	}
	return
}

// ─── Main App ──────────────────────────────────────────────────────────────

type BookApp struct {
	reader   *bufio.Reader
	catalog  *BookCatalog
}

func NewBookApp() *BookApp {
	return &BookApp{
		reader:  bufio.NewReader(os.Stdin),
		catalog: NewBookCatalog(),
	}
}

func (app *BookApp) ask(prompt string) string {
	fmt.Print(prompt)
	line, _ := app.reader.ReadString('\n')
	return strings.TrimSpace(line)
}

func (app *BookApp) askInt(prompt string, def int) int {
	for {
		ans := app.ask(prompt)
		if ans == "" {
			return def
		}
		if val, err := strconv.Atoi(ans); err == nil {
			return val
		}
		fmt.Println(c("❌ Please enter a number.", red))
	}
}

func (app *BookApp) showMenu() {
	fmt.Println("\n" + c(strings.Repeat("═", 50), cyan))
	fmt.Println(c("📚 BOOK CATALOG", bright+cyan))
	fmt.Println(c(strings.Repeat("═", 50), cyan))
	fmt.Printf("  Total books: %d\n", len(app.catalog.books))
	fmt.Printf("  Genres: %d\n", len(app.catalog.GetGenres()))
	fmt.Println(c(strings.Repeat("═", 50), cyan))
	fmt.Println("  1. 📖 List all books")
	fmt.Println("  2. ➕ Add a book")
	fmt.Println("  3. 🔍 Search books")
	fmt.Println("  4. 📊 Statistics")
	fmt.Println("  5. 🗂️  Browse by genre")
	fmt.Println("  6. ✏️  Edit a book")
	fmt.Println("  7. 🗑️  Delete a book")
	fmt.Println("  0. 🚪 Exit")
	fmt.Println(c(strings.Repeat("═", 50), cyan))
}

func (app *BookApp) listBooks(books []Book) {
	if books == nil {
		books = app.catalog.books
	}
	if len(books) == 0 {
		fmt.Println(c("No books found.", yellow))
		return
	}
	fmt.Println("\n📖 BOOKS")
	fmt.Println(c(strings.Repeat("─", 60), dim))
	for _, b := range books {
		fmt.Printf("  %s by %s (%d) – %s [ISBN: %s]\n", b.Title, b.Author, b.Year, b.Genre, b.ISBN)
	}
}

func (app *BookApp) addBook() {
	title := app.ask("Title: ")
	author := app.ask("Author: ")
	genre := app.ask("Genre: ")
	year := app.askInt("Year: ", 2000)
	isbn := app.ask("ISBN (optional): ")
	book := app.catalog.AddBook(title, author, genre, year, isbn)
	fmt.Printf(c("✅ Book '%s' added with ID %s\n", green), book.Title, book.ID)
}

func (app *BookApp) searchBooks() {
	query := app.ask("Search term (title/author/genre): ")
	genre := app.ask("Filter by genre (optional): ")
	author := app.ask("Filter by author (optional): ")
	yearMin := app.askInt("Minimum year (optional): ", 0)
	yearMax := app.askInt("Maximum year (optional): ", 9999)
	var yMin, yMax *int
	if yearMin > 0 {
		yMin = &yearMin
	}
	if yearMax < 9999 {
		yMax = &yearMax
	}
	results := app.catalog.Search(query, genre, author, yMin, yMax)
	if len(results) > 0 {
		app.listBooks(results)
	} else {
		fmt.Println(c("No books match your search.", yellow))
	}
}

func (app *BookApp) showStats() {
	total, authors, genres, minYear, maxYear := app.catalog.GetStats()
	genreMap := app.catalog.GetGenres()
	fmt.Println("\n📊 STATISTICS")
	fmt.Println(c(strings.Repeat("─", 30), dim))
	fmt.Printf("  Total Books: %d\n", total)
	fmt.Printf("  Unique Authors: %d\n", authors)
	fmt.Printf("  Genres: %d\n", genres)
	if minYear != nil && maxYear != nil {
		fmt.Printf("  Oldest Book: %d\n", *minYear)
		fmt.Printf("  Newest Book: %d\n", *maxYear)
	} else {
		fmt.Println("  Oldest Book: —")
		fmt.Println("  Newest Book: —")
	}
	if len(genreMap) > 0 {
		fmt.Println("\n📚 Books by Genre:")
		for g, count := range genreMap {
			fmt.Printf("  %s: %d\n", g, count)
		}
	}
}

func (app *BookApp) browseByGenre() {
	genres := app.catalog.GetGenres()
	if len(genres) == 0 {
		fmt.Println(c("No books yet.", yellow))
		return
	}
	genreNames := make([]string, 0, len(genres))
	for g := range genres {
		genreNames = append(genreNames, g)
	}
	fmt.Println("Select a genre:")
	for i, g := range genreNames {
		fmt.Printf("  %d. %s (%d)\n", i+1, g, genres[g])
	}
	choice := app.ask("Number: ")
	idx, err := strconv.Atoi(choice)
	if err == nil && idx >= 1 && idx <= len(genreNames) {
		selected := genreNames[idx-1]
		books := app.catalog.Search("", selected, "", nil, nil)
		app.listBooks(books)
	} else {
		fmt.Println(c("Invalid selection.", red))
	}
}

func (app *BookApp) editBook() {
	id := app.ask("Enter book ID to edit: ")
	book := app.catalog.GetBook(id)
	if book == nil {
		fmt.Println(c("Book not found.", red))
		return
	}
	fmt.Printf("Editing: %s by %s\n", book.Title, book.Author)
	title := app.ask("Title (" + book.Title + "): ")
	if title == "" {
		title = book.Title
	}
	author := app.ask("Author (" + book.Author + "): ")
	if author == "" {
		author = book.Author
	}
	genre := app.ask("Genre (" + book.Genre + "): ")
	if genre == "" {
		genre = book.Genre
	}
	year := app.askInt("Year ("+strconv.Itoa(book.Year)+"): ", book.Year)
	isbn := app.ask("ISBN (" + book.ISBN + "): ")
	if isbn == "" {
		isbn = book.ISBN
	}
	updates := map[string]interface{}{
		"title": title, "author": author, "genre": genre,
		"year": year, "isbn": isbn,
	}
	if app.catalog.UpdateBook(id, updates) {
		fmt.Println(c("✅ Book updated.", green))
	} else {
		fmt.Println(c("Failed to update.", red))
	}
}

func (app *BookApp) deleteBook() {
	id := app.ask("Enter book ID to delete: ")
	confirm := app.ask("Delete book " + id + "? (yes/no): ")
	if strings.ToLower(confirm) != "yes" {
		return
	}
	if app.catalog.DeleteBook(id) {
		fmt.Println(c("🗑️  Book deleted.", yellow))
	} else {
		fmt.Println(c("Book not found.", red))
	}
}

func (app *BookApp) run() {
	fmt.Print("\033[H\033[2J")
	fmt.Println(c("\n📚 Book Catalog – Organize Your Library", bright+cyan))
	fmt.Println(c("Manage your books by genre!", dim))

	for {
		app.showMenu()
		choice := app.ask("Your choice: ")
		switch choice {
		case "1":
			app.listBooks(nil)
		case "2":
			app.addBook()
		case "3":
			app.searchBooks()
		case "4":
			app.showStats()
		case "5":
			app.browseByGenre()
		case "6":
			app.editBook()
		case "7":
			app.deleteBook()
		case "0":
			fmt.Println(c("👋 Goodbye!", cyan))
			return
		default:
			fmt.Println(c("❌ Invalid choice.", red))
		}
		if choice != "0" {
			fmt.Print("\nPress Enter to continue...")
			app.reader.ReadString('\n')
		}
	}
}

func main() {
	app := NewBookApp()
	app.run()
}

File: book_catalog.rs
# book_catalog.rs
/**
 * 📚 Book Catalog – Organize Your Library by Genre (Rust Edition)
 * Features: add, edit, delete, search, genre stats, persistence
 * Dependencies: serde, serde_json, uuid, chrono, colored
 */

use chrono::Local;
use colored::*;
use serde::{Deserialize, Serialize};
use std::collections::{HashMap, HashSet};
use std::fs;
use std::io::{self, Write, BufRead};
use std::path::PathBuf;
use uuid::Uuid;

// ─── Data Model ─────────────────────────────────────────────────────────────

#[derive(Debug, Serialize, Deserialize, Clone)]
struct Book {
    id: String,
    title: String,
    author: String,
    genre: String,
    year: i32,
    isbn: String,
}

#[derive(Debug, Serialize, Deserialize)]
struct CatalogData {
    books: Vec<Book>,
}

// ─── Colors ──────────────────────────────────────────────────────────────────

fn c(text: &str, color: &str) -> String {
    match color {
        "green" => text.green().to_string(),
        "red" => text.red().to_string(),
        "yellow" => text.yellow().to_string(),
        "cyan" => text.cyan().to_string(),
        "bright" => text.bright().to_string(),
        "dim" => text.dimmed().to_string(),
        _ => text.to_string(),
    }
}

// ─── Catalog Manager ──────────────────────────────────────────────────────

struct BookCatalog {
    books: Vec<Book>,
    file_path: PathBuf,
}

impl BookCatalog {
    fn new() -> Self {
        let home = std::env::var("HOME").or_else(|_| std::env::var("USERPROFILE")).unwrap_or_else(|_| ".".to_string());
        let dir = PathBuf::from(home).join(".book_catalog");
        fs::create_dir_all(&dir).unwrap();
        let file_path = dir.join("books.json");
        let mut c = BookCatalog { books: Vec::new(), file_path };
        c.load();
        c
    }

    fn load(&mut self) {
        if let Ok(raw) = fs::read_to_string(&self.file_path) {
            if let Ok(data) = serde_json::from_str::<CatalogData>(&raw) {
                self.books = data.books;
                return;
            }
        }
        self.books = Vec::new();
    }

    fn save(&self) {
        let data = CatalogData { books: self.books.clone() };
        let raw = serde_json::to_string_pretty(&data).unwrap();
        let _ = fs::write(&self.file_path, raw);
    }

    fn add_book(&mut self, title: String, author: String, genre: String, year: i32, isbn: String) -> Book {
        let book = Book {
            id: Uuid::new_v4().to_string(),
            title,
            author,
            genre,
            year,
            isbn,
        };
        self.books.push(book.clone());
        self.save();
        book
    }

    fn delete_book(&mut self, id: &str) -> bool {
        let idx = self.books.iter().position(|b| b.id == id);
        if let Some(i) = idx {
            self.books.remove(i);
            self.save();
            true
        } else {
            false
        }
    }

    fn update_book(&mut self, id: &str, updates: HashMap<String, String>) -> bool {
        for book in &mut self.books {
            if book.id == id {
                if let Some(v) = updates.get("title") { book.title = v.clone(); }
                if let Some(v) = updates.get("author") { book.author = v.clone(); }
                if let Some(v) = updates.get("genre") { book.genre = v.clone(); }
                if let Some(v) = updates.get("year") {
                    if let Ok(y) = v.parse::<i32>() { book.year = y; }
                }
                if let Some(v) = updates.get("isbn") { book.isbn = v.clone(); }
                self.save();
                return true;
            }
        }
        false
    }

    fn get_book(&self, id: &str) -> Option<&Book> {
        self.books.iter().find(|b| b.id == id)
    }

    fn search(&self, query: Option<&str>, genre: Option<&str>, author: Option<&str>,
               year_min: Option<i32>, year_max: Option<i32>) -> Vec<Book> {
        let mut results = self.books.clone();
        if let Some(q) = query {
            let q = q.to_lowercase();
            results.retain(|b| {
                b.title.to_lowercase().contains(&q) ||
                b.author.to_lowercase().contains(&q) ||
                b.genre.to_lowercase().contains(&q)
            });
        }
        if let Some(g) = genre {
            let g = g.to_lowercase();
            results.retain(|b| b.genre.to_lowercase() == g);
        }
        if let Some(a) = author {
            let a = a.to_lowercase();
            results.retain(|b| b.author.to_lowercase().contains(&a));
        }
        if let Some(ymin) = year_min {
            results.retain(|b| b.year >= ymin);
        }
        if let Some(ymax) = year_max {
            results.retain(|b| b.year <= ymax);
        }
        results
    }

    fn get_genres(&self) -> HashMap<String, usize> {
        let mut map = HashMap::new();
        for b in &self.books {
            *map.entry(b.genre.clone()).or_insert(0) += 1;
        }
        map
    }

    fn get_stats(&self) -> (usize, usize, usize, Option<i32>, Option<i32>) {
        let total = self.books.len();
        let authors: HashSet<_> = self.books.iter().map(|b| &b.author).collect();
        let genres = self.get_genres().len();
        let years: Vec<i32> = self.books.iter().map(|b| b.year).collect();
        let min_year = years.iter().min().copied();
        let max_year = years.iter().max().copied();
        (total, authors.len(), genres, min_year, max_year)
    }
}

// ─── Main App ──────────────────────────────────────────────────────────────

struct BookApp {
    catalog: BookCatalog,
}

impl BookApp {
    fn new() -> Self {
        BookApp { catalog: BookCatalog::new() }
    }

    fn ask(&self, prompt: &str) -> String {
        print!("{}", prompt);
        io::stdout().flush().unwrap();
        let mut line = String::new();
        io::stdin().read_line(&mut line).unwrap();
        line.trim().to_string()
    }

    fn ask_int(&self, prompt: &str, def: i32) -> i32 {
        loop {
            let ans = self.ask(prompt);
            if ans.is_empty() { return def; }
            if let Ok(val) = ans.parse::<i32>() {
                return val;
            }
            println!("{}", c("❌ Please enter a number.", "red"));
        }
    }

    fn show_menu(&self) {
        println!("\n{}", "═".repeat(50).cyan());
        println!("{}", "📚 BOOK CATALOG".bright().cyan());
        println!("{}", "═".repeat(50).cyan());
        println!("  Total books: {}", self.catalog.books.len());
        println!("  Genres: {}", self.catalog.get_genres().len());
        println!("{}", "═".repeat(50).cyan());
        println!("  1. 📖 List all books");
        println!("  2. ➕ Add a book");
        println!("  3. 🔍 Search books");
        println!("  4. 📊 Statistics");
        println!("  5. 🗂️  Browse by genre");
        println!("  6. ✏️  Edit a book");
        println!("  7. 🗑️  Delete a book");
        println!("  0. 🚪 Exit");
        println!("{}", "═".repeat(50).cyan());
    }

    fn list_books(&self, books: Option<&[Book]>) {
        let target = match books {
            Some(b) => b,
            None => &self.catalog.books,
        };
        if target.is_empty() {
            println!("{}", c("No books found.", "yellow"));
            return;
        }
        println!("\n📖 BOOKS");
        println!("{}", "─".repeat(60).dimmed());
        for b in target {
            println!("  {} by {} ({}) – {} [ISBN: {}]", b.title, b.author, b.year, b.genre, b.isbn);
        }
    }

    fn add_book(&mut self) {
        let title = self.ask("Title: ");
        let author = self.ask("Author: ");
        let genre = self.ask("Genre: ");
        let year = self.ask_int("Year: ", 2000);
        let isbn = self.ask("ISBN (optional): ");
        let book = self.catalog.add_book(title, author, genre, year, isbn);
        println!("{}", c(&format!("✅ Book '{}' added with ID {}", book.title, book.id), "green"));
    }

    fn search_books(&self) {
        let query = self.ask("Search term (title/author/genre): ");
        let genre = self.ask("Filter by genre (optional): ");
        let author = self.ask("Filter by author (optional): ");
        let year_min = self.ask_int("Minimum year (optional): ", 0);
        let year_max = self.ask_int("Maximum year (optional): ", 9999);
        let q = if query.is_empty() { None } else { Some(query.as_str()) };
        let g = if genre.is_empty() { None } else { Some(genre.as_str()) };
        let a = if author.is_empty() { None } else { Some(author.as_str()) };
        let ymin = if year_min > 0 { Some(year_min) } else { None };
        let ymax = if year_max < 9999 { Some(year_max) } else { None };
        let results = self.catalog.search(q, g, a, ymin, ymax);
        if results.is_empty() {
            println!("{}", c("No books match your search.", "yellow"));
        } else {
            self.list_books(Some(&results));
        }
    }

    fn show_stats(&self) {
        let (total, authors, genres, min_year, max_year) = self.catalog.get_stats();
        let genre_map = self.catalog.get_genres();
        println!("\n📊 STATISTICS");
        println!("{}", "─".repeat(30).dimmed());
        println!("  Total Books: {}", total);
        println!("  Unique Authors: {}", authors);
        println!("  Genres: {}", genres);
        if let (Some(min), Some(max)) = (min_year, max_year) {
            println!("  Oldest Book: {}", min);
            println!("  Newest Book: {}", max);
        } else {
            println!("  Oldest Book: —");
            println!("  Newest Book: —");
        }
        if !genre_map.is_empty() {
            println!("\n📚 Books by Genre:");
            let mut sorted: Vec<_> = genre_map.iter().collect();
            sorted.sort_by_key(|(k, _)| *k);
            for (g, count) in sorted {
                println!("  {}: {}", g, count);
            }
        }
    }

    fn browse_by_genre(&self) {
        let genres = self.catalog.get_genres();
        if genres.is_empty() {
            println!("{}", c("No books yet.", "yellow"));
            return;
        }
        let mut genre_names: Vec<_> = genres.keys().collect();
        genre_names.sort();
        println!("Select a genre:");
        for (i, &g) in genre_names.iter().enumerate() {
            println!("  {}. {} ({})", i+1, g, genres[g]);
        }
        let choice = self.ask("Number: ");
        if let Ok(idx) = choice.parse::<usize>() {
            if idx >= 1 && idx <= genre_names.len() {
                let selected = genre_names[idx-1];
                let books = self.catalog.search(None, Some(selected), None, None, None);
                self.list_books(Some(&books));
                return;
            }
        }
        println!("{}", c("Invalid selection.", "red"));
    }

    fn edit_book(&mut self) {
        let id = self.ask("Enter book ID to edit: ");
        if let Some(book) = self.catalog.get_book(&id).cloned() {
            println!("Editing: {} by {}", book.title, book.author);
            let title = self.ask(&format!("Title ({}): ", book.title));
            let author = self.ask(&format!("Author ({}): ", book.author));
            let genre = self.ask(&format!("Genre ({}): ", book.genre));
            let year = self.ask_int(&format!("Year ({}): ", book.year), book.year);
            let isbn = self.ask(&format!("ISBN ({}): ", book.isbn));
            let mut updates = HashMap::new();
            if !title.is_empty() { updates.insert("title".to_string(), title); }
            if !author.is_empty() { updates.insert("author".to_string(), author); }
            if !genre.is_empty() { updates.insert("genre".to_string(), genre); }
            updates.insert("year".to_string(), year.to_string());
            if !isbn.is_empty() { updates.insert("isbn".to_string(), isbn); }
            if self.catalog.update_book(&id, updates) {
                println!("{}", c("✅ Book updated.", "green"));
            } else {
                println!("{}", c("Failed to update.", "red"));
            }
        } else {
            println!("{}", c("Book not found.", "red"));
        }
    }

    fn delete_book(&mut self) {
        let id = self.ask("Enter book ID to delete: ");
        let confirm = self.ask(&format!("Delete book {}? (yes/no): ", id));
        if confirm.to_lowercase() != "yes" { return; }
        if self.catalog.delete_book(&id) {
            println!("{}", c("🗑️  Book deleted.", "yellow"));
        } else {
            println!("{}", c("Book not found.", "red"));
        }
    }

    fn run(&mut self) {
        println!("{}", "\n📚 Book Catalog – Organize Your Library".bright().cyan());
        println!("{}", "Manage your books by genre!".dimmed());

        loop {
            self.show_menu();
            let choice = self.ask("Your choice: ");
            match choice.as_str() {
                "1" => self.list_books(None),
                "2" => self.add_book(),
                "3" => self.search_books(),
                "4" => self.show_stats(),
                "5" => self.browse_by_genre(),
                "6" => self.edit_book(),
                "7" => self.delete_book(),
                "0" => {
                    println!("{}", c("👋 Goodbye!", "cyan"));
                    return;
                }
                _ => println!("{}", c("❌ Invalid choice.", "red")),
            }
            if choice != "0" {
                print!("\nPress Enter to continue...");
                io::stdout().flush().unwrap();
                let mut _dummy = String::new();
                io::stdin().read_line(&mut _dummy).unwrap();
            }
        }
    }
}

// ─── Main ────────────────────────────────────────────────────────────────────

fn main() {
    let mut app = BookApp::new();
    app.run();
}
